### PR TITLE
Handle lost connections and EOF during readlines

### PR DESCRIPTION
**Pull request by warsaw**
Test cases for connection problems during readline()

Also, add some much improved debugging options for when the test suite is run
with the -E option.

Update the documentation.

Closes #61

### CODE REVIEW

**Comment by warsaw** [comment]
I grabbed the few additional non-line-length related tests from #59 although I'm not sure they don't cover cases already covered elsewhere.  I just wanted to prove to myself that this PR handles those cases too.

I feel pretty good about this one, even if we're not technically handling EOF correctly.  It does require checking for `self._connection_closed` in `SMTP.eof_received()` which is a bit ugly, but I'd like to land this branch as is, and then do a follow on branch if necessary to address that, and possibly switch over to closing the transport instead of calling `connection_lost()`.  But at least now I think we have some good test cases.

What do you think @kozzztik and @Mortal ?

File: aiosmtpd/tests/test_smtp.py [hunk 1]
```diff
@@ -790,3 +792,87 @@ def test_mail_invalid_body_param(self):
...
+
+    def test_connection_EOF_during_DATA(self):
+        with SMTP(*self.address) as client:
+            client.helo('example.com')
+            client.docmd('MAIL FROM: <anne@example.com>')
+            client.docmd('RCPT TO: <bart@example.com>')
+            client.docmd('DATA')
+            # Start sending the DATA but send an EOF before that completes,
+            # i.e. before the .\r\n
+            client.send(b'From: <anne@example.com>\0x4')
```
Comment (Mortal) [inline aiosmtpd/tests/test_smtp.py]: This should probably be `\x04`.

File: aiosmtpd/tests/test_smtp.py [hunk 2]
```diff
@@ -790,3 +792,87 @@ def test_mail_invalid_body_param(self):
...
+            # command from here will give us an exception.  In GH#62, the
+            # server just hung.
+            self.assertRaises(SMTPServerDisconnected, client.noop)
+
+    def test_connection_reset_during_command(self):
+        with SMTP(*self.address) as client:
+            client.helo('example.com')
+            # Start sending a command but reset the connection before that
+            # completes, i.e. before the \r\n
+            client.send('MAIL FROM: <anne')
```
Comment (Mortal) [inline aiosmtpd/tests/test_smtp.py]: I think this should be a bytestring for consistency with the other tests. (Does `SMTP.send` even accept a str?)

**Comment by Mortal** [comment]
Does sending `\x04` on the wire really cause an EOF? I thought signalling EOF would be simply calling `client.close()`?

**Comment by warsaw** [comment]
@Mortal I think you're right that the `\x04` isn't really necessary.  That looks like a hanger-on from a previous local commit.  Good catch.

**Comment by warsaw** [comment]
Oh, also, yes `SMTP.send()` does accept a string in the `command_encoding` which defaults to `ascii`.

**Pull request merged**